### PR TITLE
#316 Calendar original design

### PR DESCRIPTION
**Pull request by 3rdvision**
| - | - |
| --- | --- |
| Issue | https://github.com/ripe-tech/ripe-design/issues/316#issuecomment-851534607 |
| Decisions | - Add `onclick` and `keydown` emit handlers for `<input>` <br> - Add calendar single date picker component. <br> - Using native input with reactive state setter. |
| Animated GIF |  Peek 2021-05-31 16-08 |

### CODE REVIEW

**Comment by gcandal** [comment]
@3rdvision try making it so that the input and the calendar are 2 separate components, so that I can use the calendar without the input.

**Comment by gcandal** [comment]
@3rdvision try to simplify by removing the selects so this can be reviewed and merged.

File: vue/components/ui/molecules/calendar/calendar.vue [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+                        </th>
+                    </tr>
+                </thead>
+                <tbody class="table-body">
+                    <tr class="row" v-for="(week, index) in weeks" v-bind:key="index">
+                        <td
+                            class="cell"
+                            v-bind:class="circleClass(week[d - 1].day, week[d - 1].month)"
+                            v-for="d in 7"
+                            v-bind:key="d - 1"
```
Comment (gcandal) [inline vue/components/ui/molecules/calendar/calendar.vue:41]:
```suggestion
                            v-bind:key="d"
```

File: vue/components/ui/molecules/calendar/calendar.vue [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+                            {{ weekLabel }}
+                        </th>
+                    </tr>
+                </thead>
+                <tbody class="table-body">
+                    <tr class="row" v-for="(week, index) in weeks" v-bind:key="index">
+                        <td
+                            class="cell"
+                            v-bind:class="circleClass(week[d - 1].day, week[d - 1].month)"
+                            v-for="d in 7"
```
Comment (gcandal) [inline vue/components/ui/molecules/calendar/calendar.vue:40]: Try using complete and meaningful variable names.

```suggestion
                            v-for="day in 7"
```

File: vue/components/ui/molecules/calendar/calendar.vue [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+<style lang="scss" scoped>
+@import "css/variables.scss";
+
+.calendar {
+    border: 1px solid #e4e8f0;
+    border-radius: 6px 6px 6px 6px;
+    box-shadow: 0px 0px 30px rgba(21, 21, 21, 0.1);
+    display: flex;
+    flex-direction: column;
+    position: absolute;
```
Comment (gcandal) [inline vue/components/ui/molecules/calendar/calendar.vue:64]: Why is this needed?

File: vue/components/ui/molecules/calendar/calendar.vue [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+
+<script>
+export const Calendar = {
+    props: {
+        /**
+         * The value of input date in String or Date format.
+         * Example: "2020/12/31"
+         */
+        value: {
+            type: Date | String,
```
Comment (gcandal) [inline vue/components/ui/molecules/calendar/calendar.vue:151]: Don't use `Date` because it's a mutable object. This means it's prone to this kind of error:
- a = Date(10)
- v-bind:stuff="a"
- a.setMonth(3)
- v-bind doesn't trigger

File: vue/components/ui/molecules/calendar/calendar.vue [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+            default: () => ["Su", "Mo", "Tu", "We", "Th", "Fr", "Sa"]
+        }
+    },
+    data: function() {
+        return {
+            valueData: this.value,
+            calendarVisibility: true,
+            day: 1,
+            month: new Date().getMonth(),
+            year: new Date().getFullYear()
```
Comment (gcandal) [inline vue/components/ui/molecules/calendar/calendar.vue:188]: Shouldn't this be based on `value`?

File: vue/components/ui/molecules/calendar/calendar.vue [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+         */
+        weekLabels: {
+            type: Array,
+            default: () => ["Su", "Mo", "Tu", "We", "Th", "Fr", "Sa"]
+        }
+    },
+    data: function() {
+        return {
+            valueData: this.value,
+            calendarVisibility: true,
```
Comment (gcandal) [inline vue/components/ui/molecules/calendar/calendar.vue:185]: Dead code?

File: vue/components/ui/molecules/calendar/calendar.vue [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+            day: 1,
+            month: new Date().getMonth(),
+            year: new Date().getFullYear()
+        };
+    },
+    computed: {
+        circleClass() {
```
Comment (gcandal) [inline vue/components/ui/molecules/calendar/calendar.vue:192]: Method?

File: vue/components/ui/molecules/calendar/calendar.vue [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+                selected:
+                    this.selectedDay === day &&
+                    this.selectedMonth === month &&
+                    this.selectedYear === this.year
+            });
+        },
+        headerText() {
+            return `${this.monthName} ${this.year}`;
+        },
+        weekLabelsComputed() {
```
Comment (gcandal) [inline vue/components/ui/molecules/calendar/calendar.vue:204]:
```suggestion
        weekLabels() {
```

File: vue/components/ui/molecules/calendar/calendar.vue [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+        monthName() {
+            return this.monthOptions[this.month]?.label;
+        },
+        weeks() {
+            const date = new Date(this.year, this.month, 1);
+            const firstMonday = date.getDay() === 0 ? -5 : date.getDate() - date.getDay() + 1;
+            date.setDate(firstMonday);
+            const weeks = new Array(6).fill(0).map(() =>
+                new Array(7).fill(0).map(day => {
+                    day = new Date(date);
```
Comment (gcandal) [inline vue/components/ui/molecules/calendar/calendar.vue:222]: Avoid mutating variables, as it introduces cognitive load (= "wtf is 'day'" moment)

```suggestion
                new Array(7).fill(0).map(_ => {
                    const day = new Date(date);
```

File: vue/components/ui/molecules/calendar/calendar.vue [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+        selectedDay() {
+            return this.valueData?.getDate();
+        },
+        selectedMonth() {
+            return this.valueData?.getMonth();
+        },
+        selectedYear() {
+            return this.valueData?.getFullYear();
+        },
+        valueDataFormated() {
```
Comment (gcandal) [inline vue/components/ui/molecules/calendar/calendar.vue:238]: Use ripe-commons-js util function

File: vue/components/ui/molecules/input-date/input-date.vue [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+            const base = {};
+            if (this.calendarVisibility) base.visible = true;
+            return base;
+        },
+        valueDataFormated() {
+            if (!this.valueData) return;
+            const year = this.valueData.getFullYear();
+            const month = this.valueData.getMonth();
+            const day = this.valueData.getDate();
+            return `${year}-${("0" + (month + 1)).slice(-2)}-${("0" + day).slice(-2)}`;
```
[the remark this answers is not in the document]
Comment (3rdvision) [reply vue/components/ui/molecules/input-date/input-date.vue:65]: unfortunately not as it is because it needs yyyy-mm-dd format and `dateString` is returning in dd-mm-yyyy format

File: vue/components/ui/molecules/calendar/calendar.vue (hunk 3 above)
Comment (3rdvision) [reply vue/components/ui/molecules/calendar/calendar.vue:64]: this is better suited in input-date, moving...

File: vue/components/ui/molecules/calendar/calendar.vue (hunk 5 above)
Comment (3rdvision) [reply vue/components/ui/molecules/calendar/calendar.vue:188]: this is in case no initial date (`value`) was received from props but the calendar will start showing the current month and current year

imagem

File: vue/components/ui/molecules/calendar/calendar.vue (hunk 8 above)
Comment (3rdvision) [reply vue/components/ui/molecules/calendar/calendar.vue:204]: weekLabels is already a prop name

File: vue/components/ui/molecules/calendar/calendar.vue [hunk 12]
```diff
@@ -223,14 +223,18 @@ export const Calendar = {
...
     },
     watch: {
         value(value) {
-            this.setDate(value);
+            const date = new Date(parseInt(value) * 1000);
+            this.setDate(date);
         },
         valueData(value) {
-            this.$emit("update:value", value);
+            this.$emit("update:value", this.valueEpoch);
```
Comment (gcandal) [inline vue/components/ui/molecules/calendar/calendar.vue:237]: Watcher on the variable that you are using here (valueEpoch)

File: vue/components/ui/molecules/calendar/calendar.vue [hunk 13]
```diff
@@ -223,14 +223,18 @@ export const Calendar = {
         },
         selectedYear() {
             return this.valueData?.getFullYear();
+        },
+        valueEpoch() {
+            return Math.round(this.valueData.getTime() / 1000);
```
Comment (gcandal) [inline vue/components/ui/molecules/calendar/calendar.vue:228]: Is it possible to have a non-round value?

File: vue/components/ui/molecules/calendar/calendar.vue [hunk 14]
```diff
@@ -147,7 +147,7 @@ export const Calendar = {
          * Example: "2020/12/31"
          */
         value: {
-            type: String,
+            type: String | Number,
```
Comment (gcandal) [inline vue/components/ui/molecules/calendar/calendar.vue:150]: Why should we support string?

Comment (3rdvision) [reply vue/components/ui/molecules/calendar/calendar.vue:150]: I figured it could simplify extra formatting effort in case the server outputs an epoch in string format.
I can revert

File: vue/components/ui/molecules/calendar/calendar.vue (hunk 13 above)
Comment (3rdvision) [reply vue/components/ui/molecules/calendar/calendar.vue:228]: Yes, I just thought we wanted to represent dates the same way RIPE core does it which doesn't seem to include floating epoch numbers

Comment (gcandal) [reply vue/components/ui/molecules/calendar/calendar.vue:229]: Ok good, just wanted to make sure.

File: vue/components/ui/molecules/calendar/calendar.vue (hunk 14 above)
Comment (gcandal) [reply vue/components/ui/molecules/calendar/calendar.vue:150]: yea please revert, keeping interfaces as thin as possible makes them easier to maintain.

Comment (gcandal) [reply vue/components/ui/molecules/calendar/calendar.vue:150]: nvm, i know you already did

**Review by gcandal — APPROVED** [review]

**Review by joamag — APPROVED** [review]

**Review by joamag — APPROVED** [review]
